[DOC] Tutorial for semi supervised learning (#1778)
open-edge-platform/training_extensions: docs/source/guide/tutorials/advanced/semi_sl.rst

Use Semi-Supervised Learning
############################

> This tutorial provides instructions on how to use semi-supervised learning with OTX.

Reviewer: ```suggestion
This tutorial provides an example on how to use semi-supervised learning with OTX on the specific dataset.
```

> OTX now offers semi-supervised learning, which combines labeled and unlabeled data during training to improve model accuracy. Currently, this type of training is available for multi-class classification, object detection, and semantic segmentation.

Reviewer: ```suggestion
OTX now offers semi-supervised learning, which combines labeled and unlabeled data during training to improve model accuracy in case when we have small amount of annotated data. Currently, this type of training is available for multi-class classification, object detection, and semantic segmentation.
```

Semi-supervised learning will soon be available for multi-label classification and instance segmentation as well.

> If want to learn more about the algorithms used in semi-supervised learning, please refer to the explanation section below.

Reviewer: ```suggestion
If you want to learn more about the algorithms used in semi-supervised learning, please refer to the explanation section below:
```

> - :doc:`Multi-class Classification <../../explanation/algorithms/classification/multi_class_classification>`

Reviewer: Please, add there direct link to Semi-SL sections (via # or direct reference as I did in classification sections)

- :doc:`Object Detection <../../explanation/algorithms/object_detection/object_detection>`
- :doc:`Semantic Segmentation <../../explanation/algorithms/segmentation/semantic_segmentation>`

> In this tutorial, we use MobileNet-V3-large-1x model in multi-class classification to cite an example for semi-supervised learning.

Reviewer: ```suggestion
In this tutorial, we use the MobileNet-V3-large-1x model for multi-class classification to cite an example of semi-supervised learning.
```

> The process has been tested on the following configuration.

Reviewer: ```suggestion
The process has been tested on the following configuration:
```

- Ubuntu 20.04
- NVIDIA GeForce RTX 3090
- Intel(R) Core(TM) i9-10980XE
- CUDA Toolkit 11.1

.. note::
  

>   Currently, semi-supervised learning models trained in OTX cannot be exported or optimized. However, we will support for this functionality in the near future.

Reviewer: Why are they cannot be exported? The models are remains the same... We only need to omit teacher model
Is it possible to use `otx explain`, `otx demo`?

Author: It normally works just for export and POT optimization only if default template is specified. It has conflict with otx build and workspace. it does not bring compression files needed for optimization. I prefer to remain it as not supported yet till we fix it.

  This tutorial explains how to train a model in semi-supervised learning mode and how to evaluate the resulting model.

***************************
Dataset preparation
***************************

We use the same dataset, `flowers dataset <https://www.tensorflow.org/hub/tutorials/image_feature_vector#the_flowers_dataset>`_, as we do in :doc:`Classification Tutorial <../base/how_to_train/classification>`.

> Since it is assumed that we have a bunch of unlabeled data, which is not annotated but only image files exist in semi-supervised learning,

Reviewer: ```suggestion
Since it is assumed that we have additional unlabeled images,
```
unlabeled == not annotated, we can omit this for simplicity of the sentence

> we make a use of ``tests/assets/imagenet_dataset`` temporarily.

Reviewer: ```suggestion
we make a use of ``tests/assets/imagenet_dataset`` for this purpose as an example.
```

***************************
Enable via ``otx build``
***************************

> 1. To enable semi-supervsied learning via ``otx build``, we need to add arguments ``--unlabeled-data-roots`` and ``--train-type``. OTX receives the root path where unlabeled images are by ``--unlabeled-data-roots``.

Reviewer: If we omit `--train-type` and will add `--unlabeled-data-roots` only - will the `otx build` work?

Author: No it does not work. we have to put both args.

We should put the path where unlabeled data are contained. OTX also provides us ``--train-type`` to select the type of training scheme. All we have to do for that is specifying it as **SEMISUPERVISED**.

.. note::

  OTX automatically search for all image files with JPG, JPEG, and PNG formats in the root path specified using the ``--unlabeled-data-roots`` option, even if there are other file formats present. The image files which are located in sub-folders will be also collected for building unlabeled dataset.

> .. code-block::

Reviewer: please, add somewhere here, that OTX make splits for dataset automatically

  [*] Load Model Template ID: Custom_Image_Classification_MobileNet-V3-large-1x
  [*] Load Model Name: MobileNet-V3-large-1x
  [*]     - Updated: otx-workspace-CLASSIFICATION/semisl/model.py
  [*]     - Updated: otx-workspace-CLASSIFICATION/semisl/data_pipeline.py
  ...
  [*] Update data configuration file to: otx-workspace-CLASSIFICATION/data.yaml
  
  
  (otx) ...$ cd ./otx-workspace-CLASSIFICATION

2. To start training we need to call ``otx train``
command in our worspace:

.. code-block::

  (otx) ...$ otx train

In the train log, you can check that the train type is set to **SEMISUPERVISED** and related configurations are properly loaded as following:

.. code-block::

  ...
  2023-02-22 06:21:54,492 | INFO : called _init_recipe()
  2023-02-22 06:21:54,492 | INFO : train type = SEMISUPERVISED
  2023-02-22 06:21:54,492 | INFO : train type = SEMISUPERVISED - loading training_extensions/otx/recipes/stages/classification/semisl.yaml
  2023-02-22 06:21:54,500 | INFO : Replacing runner from EpochRunnerWithCancel to EpochRunnerWithCancel.
  2023-02-22 06:21:54,503 | INFO : initialized recipe = training_extensions/otx/recipes/stages/classification/semisl.yaml
  ...

> After training ends, a trained model is saved in ``models`` sub-directory in workspace ``otx-workspace-CLASSIFICATION``.

Reviewer: ```suggestion
After training ends, a trained model is saved in the ``models`` sub-directory in the workspace ``otx-workspace-CLASSIFICATION``.
```

3. In the same manner with :doc:`the normal validation <../base/how_to_train/classification#validation>`, 
we can evaluate the trained model in ``models`` folder with auto-splitted validation dataset 
and save results to ``performance.json`` by the following command:

.. code-block::

  (otx) ...$ otx eval --test-data-roots splitted_dataset/val \
                      --load-weights models/weights.pth \
                      --save-performance performance.json

***************************
Enable via ``otx train``
***************************

1. To enable semi-supervised learning directly via ``otx train``, we need to add arguments ``--unlabeled-data-roots`` and ``--algo_backend.train_type`` 

> which is one of template-specific parameters (The details are provided in :doc:`quick start guide <../../get_started/quick_start_guide/installation>`.)

Reviewer: ```suggestion
which is one of template-specific parameters (The details are provided in :doc:`quick start guide <../../get_started/quick_start_guide/cli_commands#Training>`.)
```

.. code-block::

  (otx) ...$ otx train otx/algorithms/classification/configs/mobilenet_v3_large_1_cls_incr/template.yaml \
                      --train-data-roots data/flower_photos \
                      --unlabeled-data-roots tests/assets/imagenet_dataset \
                      params --algo_backend.train_type SEMISUPERVISED

In the train log, you can check that the train type is set to **SEMISUPERVISED** and related configurations are properly loaded as following:

.. code-block::

  ...
  2023-02-22 06:21:54,492 | INFO : called _init_recipe()
  2023-02-22 06:21:54,492 | INFO : train type = SEMISUPERVISED
  2023-02-22 06:21:54,492 | INFO : train type = SEMISUPERVISED - loading training_extensions/otx/recipes/stages/classification/semisl.yaml
  2023-02-22 06:21:54,500 | INFO : Replacing runner from EpochRunnerWithCancel to EpochRunnerWithCancel.
  2023-02-22 06:21:54,503 | INFO : initialized recipe = training_extensions/otx/recipes/stages/classification/semisl.yaml
  ...

> After training ends, a trained model is saved in ``otx-workspace-CLASSIFICATION/models`` since OTX generates a workspace named ``otx-workspace-CLASSIFICATION`` in default.

Reviewer: ```suggestion
After training ends, a trained model is saved in ``./models`` by default.
```
Just to generalize all cases

Author: Well it is true though. OTX generates a workspace named otx-workspace-{task name}. I don't understand what you want to generalize.

Author: I think I got your point. Your intention was that eval should be in other section so it should be explained globally. Did I get it right? I revised it like that. Could you see again?

Reviewer: I mean, that we changed our directory to "otx-workspace-CLASSIFICATION" above and now it is ``./models``. Also, if the user will run otx train without otx build it will be saved just in ``./models`` by default as far as I remeber. So, I just wanted to omit "otx-workspace-CLASSIFICATION" in folder name since we are inside that folder. Anyway, it is just minor comment

Reviewer: I saw the changes and it looks good to me now

> 2. We can evaluate the trained model with auto-splitted validation dataset in the workspace and 

Reviewer: It is more like stand-alone section. Could you add Validation section and output example for the given command below
